Refactor gas module structure and tests

LogvinovLeon opened the pull request:
# Description

Reorganized the gas module for better maintainability and improved test quality.

### Module Organization
- Split monolithic `traits.rs` (1300+ lines) into focused submodules
- Separated `GasUnit`, `GasPrice`, `Error`, and implementation macros into dedicated files
- Moved gas meters (`BasicGasMeter`, `SlotGasMeter`, `UnlimitedGasMeter`) to `meters/` subdirectory

### Type System Improvements
- Made `GasUnit`, `GasPrice`, and `BasicGasMeter` implement `Copy`
- Removed 20+ unnecessary `.clone()` calls throughout the codebase
- Added `Copy` bound to `GasArray` trait

### Test Improvements
- Refactored tests to use table-driven/array-based parameterization for conciseness
- Added test helpers and macros (`gas!()`, assertion helpers) to reduce boilerplate
- Extracted test constants to eliminate duplication
- Split large multi-scenario tests into focused individual tests
- Added comprehensive coverage for `charge_linear_gas` edge cases
- Deleted duplicate test file (`charge_gas.rs`)

-----

- [ ] I have updated `CHANGELOG.md` with a new entry if my PR makes any breaking changes or fixes a bug. If my PR removes a feature or changes its behavior, I provide help for users on how to migrate to the new behavior.
- [ ] I have carefully reviewed all my `Cargo.toml` changes before opening the PRs. (Are all new dependencies necessary? Is any module dependency leaked into the full-node (hint: it shouldn't)?)

## Linked Issues
- Fixes # (issue, if applicable)
- Related to # (issue) 

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

Rendered docs are available at `https://sovlabs-ci-rustdoc-artifacts.us-east-1.amazonaws.com/<BRANCH_NAME>/index.html`

LogvinovLeon committed:
Reorganize gas meters into submodules

LogvinovLeon committed:
Move GasPrice and GasUnit to separate submodules

LogvinovLeon committed:
Derive promitive traits

LogvinovLeon committed:
Dedupe serde impl

LogvinovLeon committed:
Refactor our impl_macros

LogvinovLeon committed:
Move Error into submodule

LogvinovLeon committed:
Move GasUnit tests into it's submodule

LogvinovLeon committed:
Split multi-scenario test into focused individual tests

LogvinovLeon committed:
Rename tests with more descriptive names

LogvinovLeon committed:
Extract assertion helpers to reduce code duplication

LogvinovLeon committed:
Extract gas price constants to reduce duplication

LogvinovLeon committed:
Split overflow test into separate gas and funds tests

LogvinovLeon committed:
Remove unnecessary clone() call

LogvinovLeon committed:
Add comprehensive tests for charge_linear_gas method

LogvinovLeon committed:
Simplify verbose test names

LogvinovLeon committed:
Refactor gas module tests for conciseness and correctness

- Fix typos in comments and error messages
- Split large tests into focused scenarios
- Add assertion helpers to reduce boilerplate
- Extract test constants (GAS_PRICE, TEST_GAS)
- Add gas!() macro and helper functions (min, mul, add) for cleaner tests
- Refactor tests to use array-based parameterization
- Make GasUnit, GasPrice, and BasicGasMeter Copy
- Remove 20+ unnecessary .clone() calls
- Delete duplicate charge_gas.rs test file
- Move Display tests to type modules

LogvinovLeon committed:
Remove unnecessary .clone() calls on Copy types

Applied clippy --fix suggestions for clone_on_copy warnings across the codebase.
Replaced .clone() with direct copies for GasUnit, GasPrice, and related Copy types.

LogvinovLeon committed:
Fix metrics test to use &'static str for gas constant names

LogvinovLeon committed:
Remove clone() on Copy type in transaction-generator

LogvinovLeon committed:
Remove redundant field names in struct initialization

LogvinovLeon committed:
Format code

LogvinovLeon committed:
Change Gas trait to pass price by value instead of reference

Since GasPrice is now Copy, passing by value is more efficient than passing by reference.
Updated all checked_value() and value() call sites to pass GasPrice by value.

LogvinovLeon committed:
Pass GasPrice by value instead of reference

Since GasPrice is Copy, passing by value is more efficient and idiomatic.
Updated all trait methods and their implementations across the codebase:
- Gas::checked_value() and Gas::value() now accept price by value
- GetGasPrice::gas_price() now returns by value
- GasEnforcer::try_reserve_gas() now accepts gas_price by value
- Updated all function signatures and call sites throughout the codebase

This eliminates unnecessary borrows and makes the API more ergonomic.

LogvinovLeon committed:
Fix tests to pass GasPrice by value instead of reference

Updated test code to pass GasPrice by value instead of creating
unnecessary references. Changed variable bindings from:
  let gas_price = &batch_receipt.inner.gas_price;
to:
  let gas_price = batch_receipt.inner.gas_price;

This ensures consistent usage throughout the codebase.

LogvinovLeon committed:
Refactor GasArray scalar methods to consume self and return Self

Change scalar methods from mutation pattern to functional pattern:
- scalar_division(self, scalar) -> Self (was &mut self -> &mut Self)
- scalar_add(self, scalar) -> Self (was &mut self -> &mut Self)
- scalar_sub(self, scalar) -> Self (was &mut self -> &mut Self)

This follows idiomatic Rust patterns for Copy types and enables method chaining.

Also refactored GasArray comparison and combination methods to accept Copy parameters by value:
- checked_combine(&self, rhs: Self) instead of rhs: &Self
- checked_sub(&self, rhs: Self) instead of rhs: &Self
- dim_is_less_than(&self, rhs: Self) instead of rhs: &Self
- dim_is_less_or_eq(&self, rhs: Self) instead of rhs: &Self
- calculate_min(lhs: Self, rhs: Self) instead of lhs: &Self, rhs: &Self

Updated GasMeter trait to accept gas by value:
- charge_gas(&mut self, amount: Gas) instead of &Gas
- charge_linear_gas(&mut self, amount: Gas, parameter) instead of &Gas

Added #[must_use] attributes to scalar methods to prevent bugs where return values are ignored.

Fixed gas calculation bugs where scalar_division results were not being captured:
- Proving penalty calculations in test genesis configs
- Gas target calculation in ChainState
- Test setup code

Updated ~150+ call sites across the codebase to remove unnecessary & operators and capture return values.

LogvinovLeon committed:
Make GasArray methods consume self instead of &self

Change remaining GasArray methods to consume self for consistency:
- checked_combine(self, rhs: Self) instead of &self
- checked_sub(self, rhs: Self) instead of &self
- checked_scalar_product(self, scalar) instead of &self
- dim_is_less_than(self, rhs: Self) instead of &self
- dim_is_less_or_eq(self, rhs: Self) instead of &self

Since Gas types are Copy, consuming self is more idiomatic and allows
the compiler to optimize better. The methods create new values anyway,
so there's no benefit to keeping &self.

Added #[must_use] attributes to all these methods to ensure results
are not accidentally discarded.

LogvinovLeon committed:
Make Gas::checked_value and Gas::value consume self

Change Gas trait methods to consume self instead of &self:
- checked_value(self, price: Self::Price) -> Option<Amount>
- value(self, price: Self::Price) -> Amount

Added #[must_use] attributes to ensure results are not discarded.

Since Gas types are Copy, consuming self is more idiomatic and
consistent with other GasArray methods. These methods compute and
return a value, so there's no benefit to keeping &self.

LogvinovLeon committed:
Add #[must_use] to GasArray::calculate_min

This associated function returns a computed value and should have
#[must_use] to ensure the result is not accidentally discarded.

LogvinovLeon committed:
Fix linter warning: use field shorthand

LogvinovLeon committed:
Fix batch closing logic after gas method refactoring

When refactoring gas methods from mutating (&mut self -> &mut Self) to
consuming (self -> Self), the close_batch_if_nearly_full logic needed to
be updated.

The original nightly code had a bug where checked_scalar_product's result
was discarded, but because scalar_division mutated self and returned &mut Self,
the comfortable_gas_limit ended up being initial_gas_limit / 20 (5% of initial).

This was actually the correct value - we want to close batches when remaining
gas <= 5% of initial gas limit, meaning 95% of gas has been used.

The unnecessary checked_scalar_product(19) call has been removed, and a
comment was added to clarify the intended behavior.

LogvinovLeon committed:
Fix clippy warnings: remove unnecessary clones on Copy types

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/inner.rs:
```diff
@@ -513,16 +513,11 @@ where
-        let close_to_gas_limit = remaining_slot_gas.dim_is_less_or_eq(&comfortable_gas_limit);
+        // We want to close when remaining gas is at most 5% of the initial gas limit,
+        // which means we've used at least 95% of the gas.
+        let comfortable_gas_limit =
+            <S as GasSpec>::initial_gas_limit().scalar_division(COMFORTABLE_GAS_LIMIT_DIVISOR);
```
It looks like we changed the calculation here unless we also adjusted `COMFORTABLE_GAS_LIMIT_DIVISOR` to be `COMFORTABLE_GAS_LIMIT_MULTIPLIER/ COMFORTABLE_GAS_LIMIT_DIVISOR`. Is that right?

LogvinovLeon replied:
This is in fact tricky and I spent a great amount of time figuring this our. 

The original signature of: `checked_scalar_product` is: `fn checked_scalar_product(&self, scalar: $u) -> Option<Self> {`
Therefore it does not mutate self. Because we don't (didn't) assign the result of this operation - we loose it. Compiler doesn't complain as it's not `#[must_use]`. If you enable it - tests start failing as they rely on this behavior. Current behavior is correct as we take initial limit and divide by 20 getting to 5%. We later check if we have less than 5% left and close if so. I'll delete unused const

preston-evans98 replied:
Ah, I see. Maybe this would be clearer if we renamed `comfortable_gas_limit` to something `cushion_amount`? Then the comparison `remaining_slot_gas < cushion_amount` is easy to parse

LogvinovLeon replied:
I've fixed it properly by using both constants in the last commit

preston-evans98 commented on crates/module-system/sov-modules-api/Cargo.toml:
```diff
@@ -64,6 +64,7 @@ utoipa = { workspace = true, optional = true }
 utoipa-swagger-ui = { workspace = true, optional = true }
 unwrap-infallible = { workspace = true }
 nearly-linear = { workspace = true }
+arrayvec = { workspace = true, features = ["serde", "borsh"] }
```
What do we use `arrayvec` for in the new PR? A quick grep didn't show any usage...

LogvinovLeon replied:
It's a leftover. I started by replacing a String with a stack-String but later realized that we only use it in a macro that accepts 'static str. Will remove

preston-evans98 commented on crates/module-system/sov-modules-api/src/gas/tests/charge_gas.rs:
```diff
@@ -1,50 +0,0 @@
-type S = DefaultSpec<MockDaSpec, MockZkvm, MockZkvm, Native>;
-
-#[test]
-fn try_charge_gas() {
```
It looks to me like these tests were completely removed. Is that correct?

LogvinovLeon replied:
Yes. There were some tests that were duplicate of the tests that were in traits.rs. And some tests that were migrated. You can look commit by commit around the place where I touched this file to make sure I didn't loose any meaningful tests

LogvinovLeon committed:
Remove arrayvec

preston-evans98 reviewed (commented):
This change seems really nice - thanks for the effort @LogvinovLeon!

Unfortunately, the PR is really difficult to review because of the large size and the mix of API changes with simple code moves. I caught one or two (minor) issues, but with such a sprawling diff I'm afraid some may have slipped through. It looks like this history contains a bunch of smaller, more atomic changes - would it be easy to split this into chunks where each chunk does one logical thing? 

As an example...
- One PR to make the types `copy` and replace the dead clones
- (If needed) One PR to change the API from `&mut self` to `self` for the `scalar` operations. (If small, this could easily be lumped in with the previous change)
- One PR to split the large file(s) into smaller units - ideally with *zero* other changes so that the code move checker works well :)
- One PR to make the test changes)

LogvinovLeon commented:
I'm happy to split it in parts but almost each commit here can be a separate PR.
It might be easier to just go commit by commit as I tried to not mix things up within the commit.
If you wish - I'll split it into separate PRs but managing separate stacked PRs can be painful as one relies on the changes from another and I'll need to constantly be rebasing if applying fixes to the prior ones.

LogvinovLeon committed:
Refactor batch closing logic to calculate gas used instead of remaining gas

Changes the close_batch_if_nearly_full logic to:
- Calculate gas_used from initial_gas_limit - remaining_gas
- Use both COMFORTABLE_GAS_LIMIT_MULTIPLIER and DIVISOR constants (19/20 = 95%)
- Close batch when gas_used >= 95% threshold instead of remaining_gas <= 5%

This makes the logic more explicit about using both constants and clearer
about what percentage of gas has been consumed.

preston-evans98 commented:
If you want to split it into chunks that small you certainly can, but 31 PRs feels like overkill to me :)

If you can find a reviewer who's willing to sign off on this I also won't be a blocker, but I'm just not confident enough in my review to approve atm. Have to step away in a few min but if it's still open tomorrow I'll see if going thru the commits in smaller chunks makes it easier to follow!

LogvinovLeon committed:
Update lockfiles

bkolad reviewed (approved):
The refactoring looks good to me. I didn’t notice any functional changes, except that the arithmetic operations on GasArray have been made pure.
